bundles/io/org.eclipse.smarthome.io.transport.mqtt/src/main/java/org/eclipse/smarthome/io/transport/mqtt/internal/MqttBrokerConnectionServiceInstance.java

```diff
@@ -0,0 +1,118 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.core.Configuration;
+import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnection;
+import org.eclipse.smarthome.io.transport.mqtt.MqttBrokerConnectionConfig;
+import org.eclipse.smarthome.io.transport.mqtt.MqttException;
+import org.eclipse.smarthome.io.transport.mqtt.MqttService;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The user can configure multiple system Mqtt broker connections. This is realized via the OSGI service factory
+ * pattern.
+ * ESH requires a factory marker service, implemented in {@link MqttBrokerConnectionServiceInstanceMarker}.
+ * This service represents an instance of that factory and will initialize one MQTT broker connection with the given
+ * configuration and register it to the {@link MqttService}.
+ *
+ * @author David Graeff - Initial contribution
+ */
+@Component(immediate = true, configurationPolicy = ConfigurationPolicy.REQUIRE, service = MqttBrokerConnectionServiceInstance.class, configurationPid = "org.eclipse.smarthome.mqttbroker")
+@NonNullByDefault
+public class MqttBrokerConnectionServiceInstance {
+    private final Logger logger = LoggerFactory.getLogger(MqttBrokerConnectionServiceInstance.class);
+    private @Nullable MqttBrokerConnection connection;
+    private @Nullable MqttService mqttService;
+
+    @Reference
+    public void setMqttService(MqttService service) {
+        mqttService = service;
+    }
+
+    public void unsetMqttService(MqttService service) {
+        mqttService = null;
+    }
+
+    /**
+     * Create broker connections based on the service configuration. This will disconnect and
+     * discard all existing textual configured brokers.
+     */
+    @Modified
+    public void modified(@Nullable Map<String, Object> configMap) {
+        if (connection != null) {
+            connection.stop();
+        }
+
+        if (configMap == null || configMap.isEmpty() || mqttService == null) {
+            return;
+        }
+
+        // Parse configuration
+        MqttBrokerConnectionConfig config = new Configuration(configMap).as(MqttBrokerConnectionConfig.class);
+
+        try {
+            // Compute brokerID and make sure it is not empty
+            String brokerID = config.getBrokerID();
+            if (StringUtils.isBlank(brokerID)) {
+                logger.warn("Ignore invalid broker connection configuration: {}", config);
+                return;
+            }
+            assert brokerID != null;
+
+            // Add connection and make sure it succeeded
+            assert mqttService != null;
```

Comment:
True, but eclipse is issuing a warning. I guess that's because `mqttService` is a field variable and the static analyser doesn't guarantee that no concurrent thread or side effect of a method call changes the field. I would leave it for now.

Reply:
Please remove and ignore/refactor